File: order-service/OrderService/order-service.Domain/Entities/OutboxMessage.cs
using order_service.Domain.Enums;

namespace order_service.Domain.Entities;

public sealed class OutboxMessage : EntityBase
{
    public string EventType { get; private set; }
    public string Payload { get; private set; }
    public string Exchange { get; private set; }
    public string RoutingKey { get; private set; }
    public OutboxStatus Status { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime? SentAt { get; private set; }
    public int RetryCount { get; private set; }
    public string? LastError { get; private set; }  
    
    private OutboxMessage() {}

    public OutboxMessage(
        string eventType,
        string payload,
        string exchange,
        string routingKey)
    {
        EventType = eventType;
        Payload = payload;
        Exchange = exchange;
        RoutingKey = routingKey;
        Status = OutboxStatus.PENDING;
        CreatedAt = DateTime.UtcNow;
    }

    public void MarkAsSent()
    {
        Status = OutboxStatus.SENT;
        SentAt = DateTime.UtcNow;
    }

    public void MarkAsFailed(string error)
    {
        RetryCount ++;
        LastError = error;

        if (RetryCount >= 5)
            Status = OutboxStatus.Failed;
    }
}

File: fraud-service/app/domain/entities/outbox_message.py
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID, uuid4

from app.domain.entities.entity_base import EntityBase
from app.domain.enums.outbox_status import OutboxStatus

class OutboxMessage(EntityBase):
    id: UUID 
    event_type: str
    payload: str
    exchange: str
    routing_key: str
    status: OutboxStatus    
    created_at: datetime
    sent_at: Optional[datetime]
    retry_count: int
    last_error: Optional[str]

    # permite chamar métodos mutadores
    model_config = {"frozen": False} 

    @classmethod
    def create(
        cls,
        event_type: str,
        payload: str,
        exchange: str,
        routing_key: str
    ) -> "OutboxMessage":
        return cls(
            id = uuid4(),
            event_type = event_type,
            payload = payload,
            exchange = exchange,
            routing_key = routing_key,
            status = OutboxStatus.PENDING,
            created_at = datetime.now(timezone.utc),
            sent_at = None,
            retry_count = 0,
            last_error = None
        )
    
    def mark_as_sent(self) -> None:
        self.status = OutboxStatus.SENT
        self.sent_at = datetime.now(timezone.utc)
    
    def mark_as_failed(self, error: str, max_retries: int = 5) -> None:
        self.retry_count += 1
        self.last_error = error

        if self.retry_count >= max_retries:
            self.status = OutboxStatus.FAILED
